### PR TITLE
Loop invariants

### LINKED ISSUE #162: Loop invariants don't work
A regression of the compiler upgrade is that loop invariants are ignored. For example, the following verifies, but it should generate a verification error.

```rust
use prusti_contracts::*;

pub fn simple_loop() {
    let mut x = 0;
    while x < 100 {
        body_invariant!(false); // Should fail
        x += 1;
    }
}

fn main() {}
```

### PR DESCRIPTION
**Pull request by fpoli**
Fixes #162.

This PR:
* changes the desugaring of loop invariants from `let _prusti_loop_invariant = { .. }` to a closure `|| { .. }`,
* adds a `replace_multiple_places` method to `Expr`,
* simplifies the implementation of `SpecCollector`,
* refactors `spec_encoder.rs` and add a proper documentation to `encode_spec_assertion(..)`,
* removes the `namespace` field from `MirEncoder`.

To do:
* [x] Fix the remaining tests that fail due to replacements of expressions with type parameters (related: #187).
* [x] Remove the `PRUSTI_FALLBACK` feature flag and the `old_encode_expression` function, both used for debugging.
* [x] Remove the `namespace` field from `MirEncoder`.
* [x] Enable tests related to loops.
* [x] Add support for nested quantifiers

### CODE REVIEW

File: prusti-common/src/vir/ast/expr.rs [hunk 1]
```diff
@@ -1025,7 +1025,136 @@ impl Expr {
...
+                    dst,
+                    src.get_type(),
+                    dst.get_type()
+                );
+            }
+        }
+
+        struct PlaceReplacer<'a> {
+            replacements: &'a [(Expr, Expr)],
+            // FIXME: the following fields serve a grotesque hack.
```
Comment (Aurel300) [inline prusti-common/src/vir/ast/expr.rs:1050]: Maybe we should replace all of these "grotesque hack" and similar vague descriptions with links to #187 now that it exists?

File: prusti-interface/src/utils.rs [hunk 2]
```diff
@@ -297,28 +297,43 @@ pub fn has_extern_spec_attr(attrs: &[ast::Attribute]) -> bool {
...
+            use rustc_ast::token::Lit;
+            use rustc_ast::token::Token;
+            use rustc_ast::token::TokenKind;
+            use rustc_ast::tokenstream::TokenTree;
+            match &tokens.0[0].0 {
+                TokenTree::Token(Token {
+                                     kind: TokenKind::Literal(Lit { symbol, .. }),
+                                     ..
+                                 }) => {
+                    strings.push(symbol.as_str().replace("\\\"", "\""))
```
Comment (Aurel300) [inline prusti-interface/src/utils.rs:326]: This seems a bit fragile?

File: prusti-interface/src/utils.rs [hunk 3]
```diff
@@ -297,28 +297,43 @@ pub fn has_extern_spec_attr(attrs: &[ast::Attribute]) -> bool {
...
-//                 }
-//                 _ => unreachable!(),
-//             },
-//             _ => unreachable!(),
-//         },
-//         _ => unreachable!(),
-//     }
-// }
+/// Read the value stored in a single Prusti attribute (e.g. `prusti::<attr_name>="...")`.
+/// Panics if there is more than one such attribute.
```
Comment (Aurel300) [inline prusti-interface/src/utils.rs:336]: Does it?

File: prusti-interface/src/utils.rs (hunk 2 above)
Comment (fpoli) [reply prusti-interface/src/utils.rs:326]: I agree, but I don't have a quick solution. In this PR I just changed the return type from `Option<_>` to `Vec<_>`, moving the function out of `prusti-interface/src/specs/mod.rs`.

File: prusti-interface/src/utils.rs (hunk 3 above)
Comment (fpoli) [reply prusti-interface/src/utils.rs:336]: No, I forgot to update the documentation.

**Pull request merged**